tests/test_edgeql_data_migration.py

```diff
@@ -0,0 +1,1345 @@
+            }
+
+            type Derived extending Base {
+                inherited required property foo -> str;
+            }
+        """)
+        await self.con.execute("""
+            INSERT Base {
+                foo := 'base_03',
+            };
+            INSERT Derived {
+                foo := 'derived_03',
+            };
+        """)
+
+        await self._migrate(r"""
+            type Base;
+                # drop 'foo'
+
+            type Derived extending Base {
+                # completely different property
+                property foo2 -> str;
+            }
+        """)
+
+        await self.assert_query_result(
+            r"""
+                SELECT Base {
+                    __type__: {name},
+                    [IS Derived].foo2,
+                } ORDER BY .foo2;
+            """,
+            [{
+                '__type__': {'name': 'test::Base'},
+                'foo2': None,
+            }, {
+                '__type__': {'name': 'test::Derived'},
+                'foo2': None,
+            }],
+        )
+
+    async def test_edgeql_migration_04(self):
+        await self.con.execute("""
+            SET MODULE test;
+        """)
+        await self._migrate(r"""
+            type Base {
+                property foo -> str;
+            }
+
+            type Derived extending Base;
+
+            type Further extending Derived {
+                inherited required property foo -> str;
+            }
+        """)
+        await self.con.execute("""
+            INSERT Base {
+                foo := 'base_04',
+            };
+            INSERT Derived {
+                foo := 'derived_04',
+            };
+            INSERT Further {
+                foo := 'further_04',
+            };
+        """)
+
+        await self._migrate(r"""
+            type Base;
+                # drop 'foo'
+
+            type Derived extending Base;
+
+            type Further extending Derived {
+                # completely different property
+                property foo2 -> str;
+            };
+        """)
+
+        await self.assert_query_result(
+            r"""
+                SELECT Base {
+                    __type__: {name},
+                    [IS Further].foo2,
+                } ORDER BY .foo2;
+            """,
+            [{
+                '__type__': {'name': 'test::Base'},
+                'foo2': None,
+            }, {
+                '__type__': {'name': 'test::Derived'},
+                'foo2': None,
+            }, {
+                '__type__': {'name': 'test::Further'},
+                'foo2': None,
+            }],
+        )
+
+    @test.xfail('''
+        edgedb.errors.InternalServerError: column "foo" cannot be cast
+        automatically to type bigint
+
+        Ideally there should be some way of migrating regardless of
+        the potential data loss. As it stands, an extra step of column
+        dropping or renaming might be needed.
```

Comment:
We can add a `USING` clause to the `SET TYPE` operation as the mechanism to implement type changes that are not implicitly castable.  But this test is a good example of a case that requires human intervention, since the data coercion expression might be non-trivial.  Replace this with a type migration that changes to an  autocoercible type.